ouroboros-network/src/Ouroboros/Network/Tracers/OnlyForTracer.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE Rank2Types #-}
+
+-- | See 'OnlyForTracer'
+--
+-- Intended for qualified import.
+--
+-- > import           Ouroboros.Network.Tracers.OnlyForTracer (OnlyForTracer)
+-- > import qualified Ouroboros.Network.Tracers.OnlyForTracer as OnlyForTracer
+module Ouroboros.Network.Tracers.OnlyForTracer
+  ( OnlyForTracer (..)
+  , PolyRun (..)
+  , contramap
+  , mkPolyRun
+  , mkRun
+    -- * Convenience combinators
+  , contramapM
+  , lift
+  , mkTracerWith
+  , mkTracerWithPolyRun
+  , pure
+  , traceWith
+  ) where
+
+import           Prelude hiding (Applicative (..))
+import qualified Prelude
+
+import           Control.Tracer (Tracer (..))
+import qualified Control.Tracer as Tracer
+
+-- | An opaque box holding data that can only be accessed by a 'Tracer'
+--
+-- Note: In sufficient monads, the intended use can be subverted in order to
+-- leak the contained data outside of tracer invocations, but that should be
+-- caught and avoided during code review.
+newtype OnlyForTracer a = UnsafeOnlyForTracer {
+    -- | This is only safe to use _inside_ of another 'OnlyForTracer'
+    -- computation
+    unsafeRunOnlyForTracer :: a
+  }
+
+instance Functor OnlyForTracer where
+  fmap f (UnsafeOnlyForTracer a) = UnsafeOnlyForTracer (f a)
+
+instance Prelude.Applicative OnlyForTracer where
+  pure = pure
+  UnsafeOnlyForTracer f <*> UnsafeOnlyForTracer a = UnsafeOnlyForTracer (f a)
+
+-- | The only safe use of 'UnsafeOnlyForTracer'
+pure :: a -> OnlyForTracer a
+pure = UnsafeOnlyForTracer
+
+-- | The only way to eliminate the 'OnlyForTracer' structure is to let it
+-- influence a 'Tracer'
+contramap :: OnlyForTracer (a -> b) -> Tracer m b -> Tracer m a
+contramap = Tracer.contramap . unsafeRunOnlyForTracer
+
+-- | A exported wrapper only used to avoid impredicative polymorphism
```

Comment:
Add a `todo` note that we should be able to get rid of it in `ghc-9.2`.

Reply:
Done, thanks!